plugins/chain_plugin/include/eosio/chain_plugin/finality_status_object.hpp

```diff
@@ -0,0 +1,67 @@
+#pragma once
+#include <eosio/chain/types.hpp>
+#include <eosio/chain/multi_index_includes.hpp>
+#include <boost/multi_index/hashed_index.hpp>
+
+namespace bmi = boost::multi_index;
+
+namespace eosio {
+   /**
+    * @brief tracks status related to a transaction in the blockchain
+    * @ingroup object
+    *
+    * To report the status of a transaction in the system we need to track what
+    * block it is seen in as well as determine when it has failed or needs to no
+    * no longer be tracked.
+    */
+   struct finality_status_object
+   {
+      chain::transaction_id_type  trx_id;
+      fc::time_point              trx_expiry;      // if block time past trx_expiry && !block_num -> in failed list
+      fc::time_point              received;
+      chain::block_id_type        block_id;
+      chain::block_timestamp_type block_timestamp;
+      bool                        forked_out = false; // if !block_num && forked_out -> status == "forked out"
+
+      size_t memory_size() const { return sizeof(*this); }
+      bool is_in_block() const {
+         return !forked_out && block_id != chain::block_id_type{};
+      }
+   };
+
+   namespace finality_status {
+      struct by_trx_id;
+      struct by_status_expiry;
+      struct by_block_num;
+
+      using cbh = chain::block_header;
+
+      constexpr uint32_t no_block_num = 0;
+      struct block_id_comparator {
```

Comment:
I think this would be better named as `block_num_comparator`

Reply:
added a block_num() method and dropped this comparator